### PR TITLE
Optionally show disambiguating hashes in demangled Rust identifiers

### LINKED ISSUE #1754: [BUG] [Rust] all monomorphisation of a given function demangle to the same name
**Describe the bug**

When a function is monomorphized in rust, the demangled name is the same for all demangled monomorphisation of that function. If someone wants to know witch monomorphised version is called where, it needs to look are the non-demangled assembly.

**To Reproduce**

If you look in the assembly, the function `sum_print_and_skip` can be found two times. It also looks like both `caller` and `caller2` calls the same function (they don't). If you untick the `demangle` option, you can see that the full name are different for both monomorphisation.

- Rust version with `impl Trait`: https://godbolt.org/z/hr5rTM
- Rust version with generic: https://rust.godbolt.org/z/r_Amag

**Expected behavior**

It should display the generic type in the demangled name, just like what is done in C++ with template: https://rust.godbolt.org/z/YahqGR

### LINKED ISSUE #6255: [REQUEST]: Identify duplicate demangled symbols
### Is your feature request related to a problem? Please describe

In Rust, you can have multiple different symbols with the same demangled result. For example, this comes up when using closure-to-fn coercion: Rust will generate different implementations of `FnOnce::call_once` for each closure, but they all demangle to the same thing (`core::ops::function::FnOnce::call_once`).

For example:

```rust
pub fn get_adder() -> fn(u32, u32) -> u32 {
    |x, y| { x + y }
}

pub fn get_multiplier() -> fn(u32, u32) -> u32 {
    |x, y| { x * y }
}
```

This code results in [the following assembly](https://rust.godbolt.org/z/oedxrh4x6) when "demangle identifiers" is disabled:

```
_ZN4core3ops8function6FnOnce9call_once17h1d230b22b4f5ee7aE:
        lea     eax, [rdi + rsi]
        ret

_ZN4core3ops8function6FnOnce9call_once17h8b0b7f7b4815ab50E:
        mov     eax, edi
        imul    eax, esi
        ret

_ZN7example9get_adder17hcbc33e165d3b6d9dE:
        mov     eax, offset _ZN4core3ops8function6FnOnce9call_once17h1d230b22b4f5ee7aE
        ret

_ZN7example14get_multiplier17h7782378fdbbffd30E:
        mov     eax, offset _ZN4core3ops8function6FnOnce9call_once17h8b0b7f7b4815ab50E
        ret
```

However, with "demangle identifiers" enabled, the two closures seem to have the same name, and `get_adder` and `get_multiplier` appear to be identical:

```
core::ops::function::FnOnce::call_once:
        lea     eax, [rdi + rsi]
        ret

core::ops::function::FnOnce::call_once:
        mov     eax, edi
        imul    eax, esi
        ret

example::get_adder:
        mov     eax, offset core::ops::function::FnOnce::call_once
        ret

example::get_multiplier:
        mov     eax, offset core::ops::function::FnOnce::call_once
        ret
```

### Describe the solution you'd like

Perhaps append a counter at the end of functions appearing with the same demangled name. For example, something along the lines of: `core::ops::function::FnOnce::call_once`, `core::ops::function::FnOnce::call_once{1}`, `core::ops::function::FnOnce::call_once{2}`, etc.

### Describe alternatives you've considered

None

### Additional context

Not applicable

### PR DESCRIPTION
**Pull request by narpfel**
Demangled Rust identifiers under the `legacy` name mangling scheme rely on a hash to disambiguate items with the same name, such as different monomorphisations of the same function or different closures’ `call_once` methods.

In the `v0` mangling scheme, this is no longer a problem. However, configuring the demangler to show hashes in `legacy` names will include crate-id hashes in `v0` names, which are [mostly unneeded](https://rust-lang.github.io/rfcs/2603-rust-symbol-name-mangling-v0.html#appendix-a---suggested-demangling).

This PR introduces a new checkbox *Options...* → *Verbose demangling* that lets the user select if they want to include disambiguating hashes in demangled identifiers.

This checkbox is only shown for compilers that support verbose demangling, and deactivated when the *Demangle identifiers* checkbox is unchecked.

Resolves #1754.
Resolves #6255.

### CODE REVIEW

**Comment by eddyb** [comment]
> This checkbox is only shown for compilers that support verbose demangling,

FWIW, `c++filt` exposes stuff like this (it includes a Rust demangler, among other things, despite the name) with a `--no-verbose` flag (i.e. it's verbose by default), but I haven't tried figuring out what impact it has for e.g. C++ mangled symbols.

File: lib/compilers/rust.ts [hunk 1]
```diff
@@ -222,6 +223,14 @@ export class RustCompiler extends BaseCompiler {
         return options;
     }
 
+    override optionsForDemangler(filters?: ParseFiltersAndOutputOptions): string[] {
```
Comment (mattgodbolt) [inline lib/compilers/rust.ts:226]: This seems like something that could (maybe) be lifted into the baseCompiler so other compilers could take advantage of, but I see that this is very specific to `c++filt` and `--format=rust`? I see it's copy/pasted in gccrs.ts too but as it's only 6 lines I think this is probably ok.

File: lib/base-compiler.ts [hunk 2]
```diff
@@ -846,6 +846,10 @@ export class BaseCompiler implements ICompiler {
         return result;
     }
 
+    protected optionsForDemangler(filters?: ParseFiltersAndOutputOptions): string[] {
+        return [...this.compiler.demanglerArgs];
```
Comment (mattgodbolt) [inline lib/base-compiler.ts:850]: :+1:

File: static/panes/compiler.ts [hunk 3]
```diff
@@ -2741,6 +2749,14 @@ export class Compiler extends MonacoPane<monaco.editor.IStandaloneCodeEditor, Co
         // Disable demangle for compilers where we can't access it
         this.filterDemangleButton.prop('disabled', !this.compiler.supportsDemangle);
         formatFilterTitle(this.filterDemangleButton, this.filterDemangleTitle);
+
+        // Disable verbose demangling when demangling is disabled
```
Comment (mattgodbolt) [inline static/panes/compiler.ts:2753]: :+1:

**Review by mattgodbolt — APPROVED** [review]
Thanks; this seems OK to me -- as noted we could probably use tihs elsewhere too but it's a great start.

**Comment by mattgodbolt** [comment]
This is now live

File: lib/compilers/rust.ts (hunk 1 above)
Comment (narpfel) [reply lib/compilers/rust.ts:226]: I also thought about that, but it seemed to make the code more complex. If you’d like, I can make a PR to add a new field to `CompilerInfo` to set the `--no-verbose` flag in the config files? Or to deduplicate the copy/pasted code?

As eddyb said, `--no-verbose` is also available for C++ (and presumably also for the other mangling formats `c++filt` supports?). I tried to find a case in C++ where `--no-verbose` makes a difference, but I couldn’t find one.

**Comment by OfekShilon** [comment]
@narpfel  Thanks again for this. Is there a chance you can get verbose-demangling to apply to IR too? This looks like the reason for #6558.

**Comment by narpfel** [comment]
There are two problems, I think: The first (and simpler to fix) is that the regex used by the LLVM IR demangler is incorrect: 
https://github.com/compiler-explorer/compiler-explorer/blob/466346c208f1db7bcc6c9b924ceda2f8cd5f4aac/lib/demangler/llvm.ts#L35
Quoted identifiers are not recognised. ([LLVM LangRef on identifiers](https://llvm.org/docs/LangRef.html#identifiers)) See #6560.

The other is that the LLVM IR isn’t demangled through `postProcessAsm`, so this PR doesn’t apply to it. Also, I don’t think that LLVM panes know the original source language? I’d have to look closer to find a proper solution.

If I understand the code correctly, `processIrOutput` is where LLVM IR is demangled? If so, we’d probably somehow have to pass `verboseDemangling` and `this.optionsForDemangler(filters)` to the demangler around here?
https://github.com/compiler-explorer/compiler-explorer/blob/466346c208f1db7bcc6c9b924ceda2f8cd5f4aac/lib/base-compiler.ts#L1347

Also, I’m not sure though how just passing `--emit=llvm-ir` to the compiler switches from Rust demangling to LLVM IR demangling.

The Rust demangler deactivates C++ demangling by passing `--format=rust` via `demanglerArgs`, that  also gets lost for LLVM IR.

**Pull request merged**